packages/react-aria-components/src/Button.tsx

```diff
@@ -78,31 +107,133 @@ export const ButtonContext = createContext<ContextValue<ButtonContextValue, HTML
   let renderProps = useRenderProps({
     ...props,
-    values: {isHovered, isPressed, isFocused, isFocusVisible, isDisabled: props.isDisabled || false},
+    values: renderValues,
     defaultClassName: 'react-aria-Button'
   });
 
+  // an aria label will block children and their labels from being read, this is undesirable for pending state
+  let hasAriaLabel = !!buttonProps['aria-label'] || !!buttonProps['aria-labelledby'];
+  let safariDupeLabellingId = useId();
+  let buttonId = useId(buttonProps.id);
+  let contentId = useId();
+  let progressId = useId();
+
+  const isPendingAriaLiveLabel = `${hasAriaLabel ? buttonProps['aria-label'] : ''}`.trim();
+  let isPendingAriaLiveLabelledby = hasAriaLabel ? (buttonProps['aria-labelledby']?.replace(buttonId, safariDupeLabellingId) ?? safariDupeLabellingId) : `${contentId} ${safariDupeLabellingId}`.trim();
+  isPendingAriaLiveLabelledby = isPendingAriaLiveLabelledby + ' ' + progressId;
+
+  let ariaLive: 'off' | 'polite' | 'assertive' = 'polite';
+  if (isAppleDevice() && (!hasAriaLabel || isFirefox())) {
+    ariaLive = 'off';
+  }
+
+  let {textCallbackRef, progressCallbackRef} = useEnforcePendingComponents({...props, ref});
+
+  // Forcibly announce the pending state on Apple devices because otherwise it won't be announced
+  let wasPending = useRef(isPending);
+  useEffect(() => {
+    if (!wasPending.current && isFocused && isPending && isAppleDevice()) {
+      announce(isPendingAriaLiveLabelledby, 'assertive', undefined, 'ids');
+    } else if (wasPending.current && isFocused && !isPending && isAppleDevice()) {
+      announce(buttonId, 'assertive', undefined, 'ids');
+    }
+    wasPending.current = isPending;
+  }, [isPending, isFocused, isPendingAriaLiveLabelledby, buttonId]);
+
   return (
     <button
       {...filterDOMProps(props, {propNames: additionalButtonHTMLAttributes})}
       {...mergeProps(buttonProps, focusProps, hoverProps)}
       {...renderProps}
+      id={buttonId}
       ref={ref}
+      aria-label={isPending ? isPendingAriaLiveLabel : buttonProps['aria-label']}
+      aria-labelledby={isPending ? isPendingAriaLiveLabelledby : buttonProps['aria-labelledby']}
       slot={props.slot || undefined}
+      aria-disabled={isPending ? 'true' : undefined}
       data-disabled={props.isDisabled || undefined}
       data-pressed={ctx.isPressed || isPressed || undefined}
       data-hovered={isHovered || undefined}
       data-focused={isFocused || undefined}
-      data-focus-visible={isFocusVisible || undefined} />
+      data-pending={isPending || undefined}
+      data-focus-visible={isFocusVisible || undefined}>
+      <Provider
+        values={[
+          [TextContext, {id: contentId, ref: textCallbackRef}],
+          [ProgressBarContext, {id: progressId, style: {display: isPending ? undefined : 'none'}, isIndeterminate: true, ref: progressCallbackRef}]
+        ]}>
+        {renderProps.children}
+      </Provider>
+      <VisuallyHidden>
```

Comment:
Would be nice to avoid extra elements in all buttons when they aren't using pending state

Reply:
yep, now that i'm using the LiveAnnouncer, I'm trying to see if I can just not render this unless we're actually pending